### DISCUSSION TITLE
Add more control for overriding functions in GDScript

### DISCUSSION THREAD

**Discussion by shelerr** [godotengine/godot-proposals#4009] [Engine Core]
I often find myself writing code with a single base class (for example, an "enemy") and many classes inheriting from it and only overriding a few functions (something like an attack pattern).
However, it is a bit unsafe to do that since any parent function can be accidentally overridden by any of its children, which will break everything (potentially). This is especially problematic when multiple people are working on the same project.

I propose adding the following keywords:
`final` - keyword that prohibits overriding of a function
`abstract` - something like `= delete;` in C++, or `abstract` in C#. It might be a bit tricky to enforce, considering the dynamic nature of GDScript. However, I think throwing runtime error (or even warning) when attempting to instance class with deleted function is fine
`override` - same as C++/C#

As another option, it might be better to allow overriding `final` functions when using `override` (however, in that case, `final` should probably be changed to some other word).

**Comment by azur-wolve** [comment]
hi @shelerr, instead of `final`, `sealed` could be used.
Let it throw an error/warning can solve it.
